### PR TITLE
New BoringSSL context option - SSL_CTX_set1_sigalgs

### PR DESCRIPTION
**Pull request by jaymansfield**
My use case of this library had the requirement to use a different signing algorithm and certificate type (ed25519 to be specific).

Changes:

- New BoringSSL context option - SSL_CTX_set1_sigalgs (this expands on #755 )
- Allowing certificate types in certificate callback to be modified (allows the implementer to override the defaults and use any certificate type they want now)

Result:

This increases the overall flexibility of the quic implementation.

### CODE REVIEW

File: codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/BoringSSLCertificateCallback.java
Comment (normanmaurer) [inline codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/BoringSSLCertificateCallback.java:80]: why this change ?

File: codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/BoringSSLCertificateCallback.java
Comment (normanmaurer) [inline codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/BoringSSLCertificateCallback.java:69]: why this change ?

File: codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/BoringSSL.java
Comment (normanmaurer) [inline codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/BoringSSL.java:120]:
```suggestion
    }
 
```

File: codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/BoringSSL.java
Comment (normanmaurer) [inline codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/BoringSSL.java:108]: nit: Can be removed as it will throw an NPE below anyway
```suggestion
```

File: codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/BoringSSLCertificateCallback.java
Comment (normanmaurer) [inline codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/BoringSSLCertificateCallback.java:40]: why this change ?

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/BoringSSLCertificateCallback.java
Comment (jaymansfield) [reply codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/BoringSSLCertificateCallback.java:80]: This allows the person implementing this quic library to use alternate x509 certificate types if they need too. With the change, they can now override the defaults hard-coded in BoringSSLCertificateCallback. 

This change goes along with the SSL_CTX_set1_sigalgs option. The SSL_CTX_set1_sigalgs config allows the user to use different signing algorithms, and this change allows them to change the "allowed" certificate types.

File: codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/BoringSSLCertificateCallback.java
Comment (jaymansfield) [reply codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/BoringSSLCertificateCallback.java:69]: Same as previous.  Except this setting is for allowed client certificate types. The previous variable is for allowed server certificate types.

Allowing the user to add or remove from this list adds additional flexibility to this quic library. In my case, I needed to be able to use ed25519 certificates. I can now do so, without effecting other users who will keep the default settings.

File: codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/BoringSSLCertificateCallback.java
Comment (jaymansfield) [reply codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/BoringSSLCertificateCallback.java:40]: In order for a client or server to retrieve the allowed certificate type lists from the Certificate callback, the class needs to be made public rather then package protected.

File: codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/BoringSSLCertificateCallback.java
Comment (normanmaurer) [reply codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/BoringSSLCertificateCallback.java:69]: I think we should just do this via the constructor and so via a `BoringSslContextOption`

File: codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/BoringSSLCertificateCallback.java
Comment (normanmaurer) [reply codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/BoringSSLCertificateCallback.java:80]: I think we should just do this via the constructor and so via a `BoringSslContextOption`

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/BoringSSLCertificateCallback.java
Comment (jaymansfield) [reply codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/BoringSSLCertificateCallback.java]: Reverted this change

File: codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/BoringSSLCertificateCallback.java
Comment (jaymansfield) [reply codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/BoringSSLCertificateCallback.java]: Reverted this change in favour of 2 new ssl context options

File: codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/BoringSSLCertificateCallback.java
Comment (jaymansfield) [reply codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/BoringSSLCertificateCallback.java]: Reverted this change in favour of 2 new ssl context options

File: codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/BoringSSLCertificateCallback.java
Comment (normanmaurer) [inline codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/BoringSSLCertificateCallback.java:102]: Do all this in the `SslContext` implementation and just pass the `Set`directly.

**Comment by jaymansfield** [comment]
@normanmaurer Two new SSLContextOption's have been added for client and server certificate/key types

File: codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/BoringSSLCertificateCallback.java
Comment (normanmaurer) [reply codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/BoringSSLCertificateCallback.java:102]: Using the defaults when the option does not exist

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/BoringSSLCertificateCallback.java
Comment (jaymansfield) [reply codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/BoringSSLCertificateCallback.java:102]: Sounds good. Working on it.

**Comment by jaymansfield** [comment]
> LGTM @jaymansfield did you sign our icla yet ? https://netty.io/s/icla

Yes just now.

**Comment by normanmaurer** [comment]
@jaymansfield thanks a lot... will merge when the build passed.

**Comment by normanmaurer** [comment]
@jaymansfield would it be possible to add a unit test as well ?

**Comment by normanmaurer** [comment]
I think this would fit well in `QuicChannelConnectTest`

**Comment by jaymansfield** [comment]
> I think this would fit well in `QuicChannelConnectTest`

I added a very basic one for the key type change using your pre-existing RSA self signed certs in the test classes

File: codec-native-quic/src/test/java/io/netty/incubator/codec/quic/QuicChannelConnectTest.java
Comment (normanmaurer) [inline codec-native-quic/src/test/java/io/netty/incubator/codec/quic/QuicChannelConnectTest.java:748]: You want to use `ByteToMessageDecoder` as there is no guarantee that you will receive all 4 bytes at once.

File: codec-native-quic/src/test/java/io/netty/incubator/codec/quic/QuicChannelConnectTest.java
Comment (normanmaurer) [inline codec-native-quic/src/test/java/io/netty/incubator/codec/quic/QuicChannelConnectTest.java:838]:
```suggestion

```

File: codec-native-quic/src/test/java/io/netty/incubator/codec/quic/QuicChannelConnectTest.java
Comment (normanmaurer) [inline codec-native-quic/src/test/java/io/netty/incubator/codec/quic/QuicChannelConnectTest.java:869]: use `assertThrows(...)` and also use the correct `Exception` subtype that you expect.

File: codec-native-quic/src/test/java/io/netty/incubator/codec/quic/QuicChannelConnectTest.java
Comment (normanmaurer) [inline codec-native-quic/src/test/java/io/netty/incubator/codec/quic/QuicChannelConnectTest.java:865]: Just remove the listener as it will throw because of `await`.

File: codec-native-quic/src/test/java/io/netty/incubator/codec/quic/QuicChannelConnectTest.java
Comment (normanmaurer) [inline codec-native-quic/src/test/java/io/netty/incubator/codec/quic/QuicChannelConnectTest.java:862]: minimize the scope of the try block to the one that you expect to throw

File: codec-native-quic/src/test/java/io/netty/incubator/codec/quic/QuicChannelConnectTest.java
Comment (normanmaurer) [inline codec-native-quic/src/test/java/io/netty/incubator/codec/quic/QuicChannelConnectTest.java:792]: do all of this outside of a listener by awaiting the result of `createStream`. This way the assertion is also propagated to the main thread if it fails.

File: codec-native-quic/src/test/java/io/netty/incubator/codec/quic/QuicChannelConnectTest.java
Comment (normanmaurer) [inline codec-native-quic/src/test/java/io/netty/incubator/codec/quic/QuicChannelConnectTest.java:763]:
```suggestion
```

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: codec-native-quic/src/test/java/io/netty/incubator/codec/quic/QuicChannelConnectTest.java
Comment (jaymansfield) [reply codec-native-quic/src/test/java/io/netty/incubator/codec/quic/QuicChannelConnectTest.java:744]: @normanmaurer

**Comment by normanmaurer** [comment]
@jaymansfield please apply this change:

```
diff --git a/codec-native-quic/src/test/java/io/netty/incubator/codec/quic/QuicChannelConnectTest.java b/codec-native-quic/src/test/java/io/netty/incubator/codec/quic/QuicChannelConnectTest.java
index 099d244..c78436c 100644
--- a/codec-native-quic/src/test/java/io/netty/incubator/codec/quic/QuicChannelConnectTest.java
+++ b/codec-native-quic/src/test/java/io/netty/incubator/codec/quic/QuicChannelConnectTest.java
@@ -27,6 +27,7 @@ import io.netty.channel.ChannelOutboundHandlerAdapter;
 import io.netty.channel.ChannelPromise;
 import io.netty.channel.ConnectTimeoutException;
 import io.netty.channel.socket.ChannelInputShutdownEvent;
+import io.netty.handler.codec.ByteToMessageDecoder;
 import io.netty.handler.ssl.ClientAuth;
 import io.netty.handler.ssl.SniCompletionEvent;
 import io.netty.handler.ssl.SslHandshakeCompletionEvent;
@@ -69,8 +70,18 @@ import java.security.cert.CertificateException;
 import java.security.cert.X509Certificate;
 import java.security.spec.MGF1ParameterSpec;
 import java.security.spec.PSSParameterSpec;
-import java.util.*;
-import java.util.concurrent.*;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executor;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.Consumer;
@@ -741,24 +752,16 @@ public class QuicChannelConnectTest extends AbstractQuicTest {
                     public boolean isSharable() {
                         return true;
                     }
-                }, new ChannelInboundHandlerAdapter() {
+                }, new ByteToMessageDecoder() {
                     @Override
-                    public boolean isSharable() {
-                        return true;
-                    }
-
-                    @Override
-                    public void channelRead(ChannelHandlerContext ctx, Object msg) {
-                        ByteBuf buffer = (ByteBuf) msg;
-                        try {
-                            assertEquals(4, buffer.readableBytes());
-                            assertEquals(5, buffer.readInt());
-                            readLatch.countDown();
-
-                            ctx.close();
-                        } finally {
-                            buffer.release();
+                    protected void decode(ChannelHandlerContext ctx, ByteBuf msg, List<Object> out) throws Exception {
+                        if (msg.readableBytes() < 4) {
+                            return;
                         }
+                        assertEquals(5, msg.readInt());
+                        readLatch.countDown();
+
+                        ctx.close();
                     }
                 });
 

```

File: codec-native-quic/src/test/java/io/netty/incubator/codec/quic/QuicChannelConnectTest.java
Comment (normanmaurer) [inline codec-native-quic/src/test/java/io/netty/incubator/codec/quic/QuicChannelConnectTest.java:90]: please revert this one... We don't do wildcard imports

File: codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/BoringSSLCertificateCallback.java
Comment (normanmaurer) [inline codec-classes-quic/src/main/java/io/netty/incubator/codec/quic/BoringSSLCertificateCallback.java:19]: Unused 

```suggestion
```

**Review by normanmaurer — CHANGES_REQUESTED** [review]
Almost there... two more things. After this we can merge

**Review by normanmaurer — APPROVED** [review]

**Pull request merged**
